core/src/main/java/io/undertow/server/protocol/ajp/AjpServerResponseConduit.java

```diff
@@ -266,6 +268,40 @@ protected void queueCloseFrames() {
         queueFrame(null, buffer);
     }
 
+    private void queueRemainingBytes(final ByteBuffer src, final ByteBuffer[] buffers) {
+        List<PooledByteBuffer> pools = new ArrayList<>(4);
+
+        try {
+            PooledByteBuffer newPooledBuffer = pool.allocate();
+            pools.add(newPooledBuffer);
+            while (src.remaining() > newPooledBuffer.getBuffer().remaining()) {
+                ByteBuffer dupSrc = src.duplicate();
+                dupSrc.limit(dupSrc.position() + newPooledBuffer.getBuffer().remaining());
+                newPooledBuffer.getBuffer().put(dupSrc);
+                src.position(dupSrc.position());
+                newPooledBuffer.getBuffer().flip();
+                newPooledBuffer = pool.allocate();
+                pools.add(newPooledBuffer);
+            }
+            newPooledBuffer.getBuffer().put(src);
+            newPooledBuffer.getBuffer().flip();
+
+            ByteBuffer[] savedBuffers = new ByteBuffer[pools.size() + 2];
+            int i = 0;
+            savedBuffers[i++] = buffers[0];
+            for (PooledByteBuffer p : pools) {
+                savedBuffers[i++] = p.getBuffer();
+            }
+            savedBuffers[i] = buffers[2];
+            queueFrame(new PooledBuffersFrameCallback(pools.toArray(new PooledByteBuffer[0])), savedBuffers);
+        } catch (RuntimeException | Error e) {
```

Comment:
better use a finally block

Reply:
Hi @fl4via! 

I think the problem is not at this point. It's what I tried to explain yesterday in the meeting. The buffers here are saved because writes returned 0 (no more writing) and they need to be queued to send them later (when writes are called again). So a finally block to close them is not valid, they should be created now and closed when fully sent. My feeling is that the request is somehow hanging and still on-going in the backend server but the proxy timed out (note the test takes more than 300s, and it usually is around 20s). Other thing is why the test is a success, which is very strange. But it can be a ConnectionClosedException which is ignored [here](https://github.com/undertow-io/undertow/blob/master/servlet/src/test/java/io/undertow/servlet/test/streams/ServletOutputStreamTestCase.java#L271-L275). But I'm not sure about this part.

In summary, I think we have two separate issues involving AjpServerResponseConduit, and this PR just fixes the easy one. We need to investigate what happens later and why the buffer is not closed when sent. It can be because it hangs (maybe related to UNDERTOW-1945 or UNDERTOW-1948) or a missing condition on close (but I have not seen any obvious mistake yet).

What do you think about creating a followup JIRA and at least merging this fix?

Reply:
@rmartinc thanks for the explanation. That sounds fine.

Reply:
@rmartinc  I'm not sure if the failure I saw a couple of days ago was the same. I have just created https://issues.redhat.com/browse/UNDERTOW-2054 to track the Ssl buffer leak errors on Windows proxy-https tests. But I have the impression that, when we were discussing this, we were talking about a buffer created by AjpServerResponseConduit, can yo confirm? As soon as I see that failure again, I'll create the Jira.

Reply:
You can disregard this comment, I found the issue again and I saw your comments in the Jira, here: https://issues.redhat.com/browse/UNDERTOW-2028?focusedCommentId=20027058&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-20027058
Let's proceed with UNDERTOW-1945 as it seems it is the same failure indeed. Thanks for having a look at this!